python/ray/serve/master.py

```diff
@@ -479,6 +493,47 @@ async def create_endpoint(self, route, endpoint, methods):
             self._checkpoint()
             await self.http_proxy.set_route_table.remote(self.routes)
 
+    async def delete_endpoint(self, endpoint):
+        """Delete the specified endpoint.
+
+        Does not modify any corresponding backends.
+        """
+        logger.info("Deleting endpoint '{}'".format(endpoint))
+        async with self.write_lock:
+            # This method must be idempotent. We should validate that the
+            # specified endpoint exists on the client.
+            if endpoint not in self.traffic_policies:
+                logger.info("Endpoint '{}' doesn't exist".format(endpoint))
+                return
+
+            # Remove the traffic policy entry.
+            del self.traffic_policies[endpoint]
+
+            for route, (route_endpoint, _) in self.routes.items():
+                if route_endpoint == endpoint:
+                    route_to_delete = route
+                    break
+            else:
+                # This should never happen, we either add to or delete from
+                # both self.traffic_policies and self.routes.
+                assert False, "No route found for endpoint '{}'".format(
+                    endpoint)
+
+            # Remove the routing entry.
+            del self.routes[route_to_delete]
```

Comment:
```suggestion
            for route, (route_endpoint, _) in self.routes.items():
                if route_endpoint == endpoint:
                    route_to_delete = route
                    del self.routes[route_to_delete]
```
Just making this no op if endpoint doesn't exist in self.routes?

Reply:
This is an internal invariant check, see the comment

Reply:
The invariant might not be true for no route endpoints?